tools/syz-declextract/run.go

```diff
@@ -62,35 +64,94 @@ func main() {
 	syscallNames := readSyscallNames(filepath.Join(*kernelDir, "arch")) // some syscalls have different names and entry
 	// points and thus need to be renamed.
 	// e.g. SYSCALL_DEFINE1(setuid16, old_uid_t, uid) is referred to in the .tbl file with setuid.
+
+	eh := ast.LoggingHandler
 	for range cmds {
 		out := <-outputs
 		if out.stderr != "" {
 			tool.Failf("%s", out.stderr)
 		}
-		for _, line := range strings.Split(out.stdout, "\n") {
-			if line == "" {
-				continue
+		parse := ast.Parse([]byte(out.stdout), "", eh)
+		if parse == nil {
+			tool.Failf("Parsing Error\n")
+		}
+		for _, node := range parse.Nodes {
+			switch node.(type) {
+			case *ast.Call:
+				syscalls = append(syscalls, renameSyscall(node.(*ast.Call), syscallNames)...)
+			case *ast.Struct:
+				netlinks = append(netlinks, node.(*ast.Struct))
+			case *ast.Include:
+				includes = append(includes, node.(*ast.Include))
 			}
-			allOut = append(allOut, renameSyscall(line, syscallNames)...)
 		}
 	}
+
 	close(files)
-	writeOutput(allOut, *outFile)
+	writeOutput(includes, syscalls, netlinks, *outFile)
 }
 
-func writeOutput(allOut []string, outFile string) {
-	slices.Sort(allOut)
-	allOut = slices.CompactFunc(allOut, func(a string, b string) bool {
-		// We only compare the part before "$" for cases where the same system call has different parameter names,
+func writeOutput(includes []*ast.Include, syscalls []*ast.Call, netlinks []*ast.Struct, outFile string) {
+
+	// includes = []*ast.Node(includes)
+
+	slices.SortFunc(includes, func(a, b *ast.Include) int {
+		return strings.Compare(a.File.Value, b.File.Value)
+	})
+	includes = slices.CompactFunc(includes, func(a, b *ast.Include) bool {
+		return a.File.Value == b.File.Value
+	})
+
+	slices.SortFunc(syscalls, func(a, b *ast.Call) int {
+		return strings.Compare(a.Name.Name, b.Name.Name)
+	})
+	syscalls = slices.CompactFunc(syscalls, func(a, b *ast.Call) bool {
+		// We only compare the the system call names for cases where the same system call has different parameter names,
 		// but share the same syzkaller type. NOTE:Change when we have better type extraction.
-		return strings.Split(a, "$")[0] == strings.Split(b, "$")[0]
+		return a.Name.Name == b.Name.Name
 	})
+	slices.SortFunc(netlinks, func(a, b *ast.Struct) int {
+		return strings.Compare(a.Name.Name, b.Name.Name)
+	})
+
+	autoGeneratedNotice := "# Code generated by syz-declextract. DO NOT EDIT."
+	mmap2 := "_ = __NR_mmap2\n"
+	eh := ast.LoggingHandler
+	desc := ast.Parse([]byte(autoGeneratedNotice), "", eh)
+	for _, node := range includes {
+		desc.Nodes = append(desc.Nodes, node)
+	}
+	for _, node := range syscalls {
+		desc.Nodes = append(desc.Nodes, node)
+	}
+	desc.Nodes = append(desc.Nodes, ast.Parse([]byte(mmap2), "", eh).Nodes...)
+	var netlinkNames []string
+	for _, node := range netlinks {
+		desc.Nodes = append(desc.Nodes, node)
+		netlinkNames = append(netlinkNames,
+			"\tpolicy"+strconv.Itoa(len(netlinkNames))+" msghdr_auto["+node.Name.Name+"]\n")
```

Comment:
Will be a bit simpler if you add index to the range loop and use `i` here:
```
for i, node := range netlinks {
```

also for a complex string that is formed by string concatenation of random expression, it's hard to understand what will be the resulting string.
Use of string formatting makes it easier to understand what's the resulting string since we have the string and the data separated:
```
fmt.Sprintf("\tpolicy%v msghdr_auto[%v]\n", i, node.Name.Name)
```

Reply:
Done! Much clearer!